allennlp/modules/seq2seq_encoders/multi_head_self_attention.py

```diff
@@ -101,33 +93,33 @@ def forward(self,  # pylint: disable=arguments-differ
         """
         num_heads = self._num_heads
 
-        batch_size, timesteps, hidden_dim = inputs.size()
+        batch_size, timesteps, _ = inputs.size()
         if mask is None:
             mask = Variable(inputs.data.new(batch_size, timesteps).fill_(1.0))
 
-        # Treat the queries, keys and values each as a ``num_heads`` size batch.
-        # shape (num_heads, batch_size * timesteps, hidden_dim)
-        inputs_per_head = inputs.repeat(num_heads, 1, 1).view(num_heads,
-                                                              batch_size * timesteps,
-                                                              hidden_dim)
-        # Do the projections for all the heads at once.
-        # Then reshape the result as though it had a
-        # (num_heads * batch_size) sized batch.
-        queries_per_head = torch.bmm(inputs_per_head, self._query_projections)
-        # shape (num_heads * batch_size, timesteps, attention_dim)
-        queries_per_head = queries_per_head.view(num_heads * batch_size,
-                                                 timesteps,
-                                                 self._attention_dim)
-
-        keys_per_head = torch.bmm(inputs_per_head, self._key_projections)
-        # shape (num_heads * batch_size, timesteps, attention_dim)
-        keys_per_head = keys_per_head.view(num_heads * batch_size,
-                                           timesteps,
-                                           self._attention_dim)
-
-        values_per_head = torch.bmm(inputs_per_head, self._value_projections)
-        # shape (num_heads * batch_size, timesteps, attention_dim)
-        values_per_head = values_per_head.view(num_heads * batch_size, timesteps, self._values_dim)
+        # Shape (batch_size, timesteps, 2 * attention_dim + values_dim)
```

Comment:
@matt-gardner can you think of a better way to do this so I can avoid doing these multiple calls to `view` and `contiguous`?